utilities/transactions/lock/range/range_tree/range_tree_lock_manager.h

```diff
@@ -0,0 +1,128 @@
+//
+// Range Lock Manager that uses PerconaFT's locktree library
+//
+#pragma once
+#ifndef ROCKSDB_LITE
+#ifndef OS_WIN
+
+// For DeadlockInfoBuffer:
+#include "util/thread_local.h"
+#include "utilities/transactions/lock/point/point_lock_manager.h"
+#include "utilities/transactions/lock/range/range_lock_manager.h"
+
+// Lock Tree library:
+#include "lib/locktree/lock_request.h"
+#include "lib/locktree/locktree.h"
+#include "range_tree_lock_tracker.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+using namespace toku;
+
+typedef DeadlockInfoBufferTempl<RangeDeadlockPath> RangeDeadlockInfoBuffer;
+
+/*
+  A Range Lock Manager that uses PerconaFT's locktree library
+*/
+class RangeTreeLockManager : public RangeLockManagerBase,
+                             public RangeLockManagerHandle {
+ public:
+  LockManager* getLockManager() override { return this; }
+
+  void AddColumnFamily(const ColumnFamilyHandle* cfh) override;
+  void RemoveColumnFamily(const ColumnFamilyHandle* cfh) override;
+
+  void Resize(uint32_t) override;
+  std::vector<DeadlockPath> GetDeadlockInfoBuffer() override;
+
+  std::vector<RangeDeadlockPath> GetRangeDeadlockInfoBuffer() override;
+  void SetRangeDeadlockInfoBufferSize(uint32_t target_size) override;
+
+  // Get a lock on a range
+  //  @note only exclusive locks are currently supported (requesting a
+  //  non-exclusive lock will get an exclusive one)
```

Comment:
what about returning an error status when requesting a non-exclusive one?

Reply:
I think returning a non-exclusive lock is better.  It still meets the call contract  (and with lock escalation, there is no guarantee that the lock is obtained on the exact range requested. A broader lock can be acquired, instead).